### PR TITLE
mgr/restful: support for http requests

### PR DESCRIPTION
**Pull request by hyuanchun**
Support ceph config set mgr mgr/restful/https_mode false/true to control HTTP/HTTPS access.

Fixes: https://tracker.ceph.com/issues/58064

Signed-off-by: huangyuanchun <huangyuanchun_yewu@cmss.chinamobile.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by hyuanchun** [comment]
@pjjw @nSedrickm @Pegonzal @pereman2

File: src/pybind/mgr/restful/module.py [hunk 1]
```diff
@@ -314,8 +316,12 @@ def _serve(self):
...
         # Publish the URI that others may use to access the service we're
         # about to start serving
-        addr = self.get_mgr_ip() if server_addr == "::" else server_addr
-        self.set_uri(build_url(scheme='https', host=addr, port=server_port, path='/'))
+        if self.https_mode:
+            addr = self.get_mgr_ip() if server_addr == "::" else server_addr
+            self.set_uri(build_url(scheme='https', host=addr, port=server_port, path='/'))
+        else:
+            addr = self.get_mgr_ip() if server_addr == "::" else server_addr
+            self.set_uri(build_url(scheme='http', host=addr, port=server_port, path='/'))
```
Comment (pereman2) [inline src/pybind/mgr/restful/module.py:324]:
```suggestion
        addr = self.get_mgr_ip() if server_addr == "::" else server_addr
        if self.https_mode:
            self.set_uri(build_url(scheme='https', host=addr, port=server_port, path='/'))
        else:
            self.set_uri(build_url(scheme='http', host=addr, port=server_port, path='/'))
```

**Review by pereman2 — APPROVED** [review]

**Comment by pereman2** [comment]
What are you usually using the restful module for?

File: src/pybind/mgr/restful/module.py (hunk 1 above)
Comment (hyuanchun) [reply src/pybind/mgr/restful/module.py]: done

**Comment by hyuanchun** [comment]
> What are you usually using the restful module for?
Manage the ceph cluster

**Comment by hyuanchun** [comment]
Any other suggestions?  @pereman2 @avanthakkar @nizamial09

**Comment by tchaikov** [comment]
the thing is, we do have a plan to deprecate mgr/restful. see https://tracker.ceph.com/issues/47066 .

**Comment by hyuanchun** [comment]
> the thing is, we do have a plan to deprecate mgr/restful.

I wonder how long it will take to complete the project. For the time being, I think it can be merged.

**Comment by hyuanchun** [comment]
> > the thing is, we do have a plan to deprecate mgr/restful.
> 
> I wonder how long it will take to complete the project. For the time being, I think it can be merged.

@tchaikov

**Comment by hyuanchun** [comment]
Can you merge it? Any other suggestions? @pereman2

**Comment by tchaikov** [comment]
> > the thing is, we do have a plan to deprecate mgr/restful.
> 
> I wonder how long it will take to complete the project. For the time being, I think it can be merged.

my understanding is, the restful module is deprecated, and the reason we have it is only for backward compatibility. no new features should be added.

**Comment by tchaikov** [comment]
@pereman2 that's my understanding back when @epuertat sent a mail asking for input regarding the removal of this mgr module. i think you guys have the final call.

**Comment by hyuanchun** [comment]
> @tchaikov If we planned to deprecate and not add any more features let's stick that way. @hyuanchun Is there a reason to not use the Dashboard API?
We'll replace it with prometheus and grafana; We don't use Dashboard.

**Pull request closed without merging**